### ISSUE TITLE
Data Source Lookup - TF Crash

### ISSUE THREAD

**Issue by huseyinham** [terraform-providers/terraform-provider-aws#10469]
### Terraform Version
```
Terraform v0.11.10
+ provider.aws v2.31.0
+ provider.template v2.1.2
```

### Terraform Configuration Files
Main components that cause the error:

```hcl
resource "aws_lambda_function" "lambda_trigger_file_drop" {
  ...
  vpc_config {
	  subnet_ids = ["${data.aws_subnet_ids.subnet_ids.ids}"]
	  security_group_ids = ["${data.aws_security_group.lambda_sg.id}"]
  }
}

data "aws_security_group" "lambda_sg" {
  name = "testing-lambda-sg"
}

data "aws_subnet_ids" "subnet_ids" {
  vpc_id = "${data.aws_vpcs.vpcs.ids}"
}

data "aws_vpcs" "vpcs" {
    filter {
        name = "tag:Name"
        values = ["*-${terraform.env}"]
    }
}
```

### Debug Output
https://gist.github.com/huseyinham/061d7d830418c43a72a4c1c4dd87479f

### Panic Output
```
2019-10-11T10:55:36.538+0100 [DEBUG] plugin.terraform-provider-aws_v2.31.0_x4: panic: Error reading level config: '' expected type 'string', got unconvertible type '[]interface {}'
```

### Expected Behavior

Lambda resource builds and attaches Subnet IDs and Security Group IDs.

### Actual Behavior

Terraform Crash.

I have found other issues which state similar things happening. However most have been slightly different bugs and on slightly different TF versions etc. I imagine the answer will be that I need to upgrade TF version (as the version is quite old now), but I suppose I just want some else's opinion before I go through the process of upgrading.

**Comment by huseyinham** [comment]
Just to add, a plan works if I change the following to be hard-coded instead:

```
data "aws_subnet_ids" "subnet_ids" {
  vpc_id = "myvpcid-123456"
}
```

Can I assume that you can't do a datasource lookup inside another datasource?

**Comment by bflad** [comment]
Hi @huseyinham 👋 Sorry you are running into this frustrating issue.

> I imagine the answer will be that I need to upgrade TF version (as the version is quite old now), but I suppose I just want some else's opinion before I go through the process of upgrading.

Your hunch here is actually correct in this case -- Terraform 0.11 and earlier did not always notify operators of configuration errors such as this. In Terraform 0.12 and later, type mismatches are surfaced much better. In this case:

```hcl
# Example configuration in Terraform 0.12 syntax
# Using "Default VPC" so configuration passes in my testing account
terraform {
  required_providers {
    aws = "2.32.0"
  }

  required_version = "0.12.10"
}

provider "aws" {
  region = "us-east-1"
}

data "aws_vpcs" "vpcs" {
  filter {
    name   = "tag:Name"
    values = ["Default VPC"]
  }
}

data "aws_subnet_ids" "subnet_ids" {
  vpc_id = data.aws_vpcs.vpcs.ids
}

```

```console
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_vpcs.vpcs: Refreshing state...

Error: Incorrect attribute value type

  on main.tf line 21, in data "aws_subnet_ids" "subnet_ids":
  21:   vpc_id = data.aws_vpcs.vpcs.ids

Inappropriate value for attribute "vpc_id": string required.
```

As Terraform 0.12 can point out, the value returned from the `aws_vpcs` data source's `ids` attribute is a list, instead of the string expected by the `aws_subnet_ids` data source's `vpc_id` argument. To resolve this, you will likely want to use [`count`](https://www.terraform.io/docs/configuration-0-11/resources.html#count) in Terraform 0.11, (likely more preferable) [`for_each`](https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings) in Terraform 0.12.6 or later, or switch to using the [`aws_vpc` data source](https://www.terraform.io/docs/providers/aws/d/vpc.html) (singular not plural) instead of `aws_vpcs` (plural). Which to choose will be dependent on your use case and in general the maintainers here cannot provide much additional guidance on this subject.

> Can I assume that you can't do a datasource lookup inside another datasource?

You can and its a fairly common pattern if you are managing your network in separate Terraform configurations/modules. 😄 The only limitations are generally to not use a data source lookup on a resource that the same configuration creates and [to avoid `depends_on` usage with data sources](https://www.terraform.io/docs/configuration/data-sources.html#data-resource-dependencies).

---

Since this particular crash is resolved by upgrading Terraform CLI to v0.12.0 or later, I'm going to close this issue. If you're looking for general assistance beyond what I provided above, please note that we use GitHub issues in this repository for tracking bugs and enhancements with the Terraform AWS Provider codebase rather than for questions. While we may be able to help with certain simple problems here it's generally better to use the [community forums](https://discuss.hashicorp.com/c/terraform-providers) where there are far more people ready to help, whereas the GitHub issues here are generally monitored only by a few maintainers and dedicated community members interested in code development of the Terraform AWS Provider itself. Thanks!
